sebs/fission/fission.py

```diff
@@ -0,0 +1,373 @@
+        minioConfig.close()
+        scriptPath = os.path.join(directory, "build.sh")
+        self.shouldCallBuilder = True
+        f = open(scriptPath, "w+")
+        f.write(
+            "#!/bin/sh\npip3 install -r ${SRC_PKG}/requirements.txt -t \
+${SRC_PKG} && cp -r ${SRC_PKG} ${DEPLOY_PKG}"
+        )
+        f.close()
+        subprocess.run(["chmod", "+x", scriptPath])
+        for file in os.listdir(directory):
+            if file not in package_config:
+                file = os.path.join(directory, file)
+                shutil.move(file, function_dir)
+        os.chdir(directory)
+        subprocess.run(
+            "zip -r {}.zip ./".format(benchmark.benchmark).split(),
+            stdout=subprocess.DEVNULL,
+        )
+        benchmark_archive = "{}.zip".format(os.path.join(directory, benchmark.benchmark))
+        logging.info("Created {} archive".format(benchmark_archive))
+        bytes_size = os.path.getsize(benchmark_archive)
+        return benchmark_archive, bytes_size
+
+    def update_function(self, name: str, env_name: str, code_path: str):
+        self.create_function(name, env_name, code_path)
+
+    def create_env_if_needed(self, name: str, image: str, builder: str):
+        self.envName = name
+        try:
+            fission_env_list = subprocess.run(
+                "fission env list ".split(),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.DEVNULL,
+            )
+            subprocess.run(
+                f"grep {name}".split(),
+                check=True,
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.DEVNULL,
+                input=fission_env_list.stdout,
+            )
+            logging.info(f"Env {name} already exist")
+        except subprocess.CalledProcessError:
+            logging.info(f'Creating env for {name} using image "{image}".')
+            try:
+                subprocess.run(
+                    f"fission env create --name {name} --image {image} \
+                    --builder {builder}".split(),
+                    check=True,
+                    stdout=subprocess.DEVNULL,
+                )
+            except subprocess.CalledProcessError:
+                logging.info(f"Creating env {name} failed. Retrying...")
+                sleep(10)
+                try:
+                    subprocess.run(
+                        f"fission env create --name {name} --image {image} \
+                        --builder {builder}".split(),
+                        check=True,
+                        stdout=subprocess.DEVNULL,
+                    )
+                except subprocess.CalledProcessError:
+                    self.storage.storage_container.kill()
+                    logging.info("Minio stopped")
+                    self.initialize()
+                    self.create_env_if_needed(name, image, builder)
+
+    def create_function(self, name: str, env_name: str, path: str):
```

Comment:
I'm not sure what is the relation between package and function in Fission?

Reply:
packages hold everything that a function needs like the code package, function are wrappers outside of the package where we provide that package the function runs, what is the entry point, memoery and timeout configs etc